### PR TITLE
Integrate Ansible

### PR DESCRIPTION
**Pull request by glenn-sorrentino**
- Added setup.yml
- Added common.yml
- Added encrypted secrets/secrets.yml
- Installed Ansible locally and tested deploying.
   - `ansible-playbook -i <IP>, setup.yml --ask-vault-pass -e ansible_ssh_private_key_file=~/.ssh/dev -e ansible_user=root`
    - [Gist of last install](https://gist.github.com/glenn-sorrentino/4f2cc83825d06c63afdf79d43d0dc5e8). Hitting a persistent issue where installing dependencies with Poetry fails. 
       - [When manually attempting installation](https://gist.github.com/glenn-sorrentino/4b656e67ea664128b3f53d2ea084c129), I activated the environment and executed the installation command in the playbook and it succeeds.

### CODE REVIEW

**Comment by glenn-sorrentino** [comment]
> * Added setup.yml
> 
>     * Added common.yml
> 
>     * Added encrypted secrets/secrets.yml
> 
>     * Installed Ansible locally and tested deploying.
>       
>       * `ansible-playbook -i <IP>, setup.yml --ask-vault-pass -e ansible_ssh_private_key_file=~/.ssh/dev -e ansible_user=root`
>       * [Gist of last install](https://gist.github.com/glenn-sorrentino/4f2cc83825d06c63afdf79d43d0dc5e8). Hitting a persistent issue where installing dependencies with Poetry fails.
>         
>         * [When manually attempting installation](https://gist.github.com/glenn-sorrentino/4b656e67ea664128b3f53d2ea084c129), I activated the environment and executed the installation command in the playbook and it succeeds.

Resolved by activating the venv, then calling poetry by it full path:

```
- name: Install dependencies with Poetry
      shell: |
        cd /var/www/html/hushline && source /var/www/html/hushline/venv/bin/activate && /root/.local/bin/poetry install
      args:
        executable: /bin/bash
```

File: install.sh
Comment (brassy-endomorph) [inline install.sh:1]: Ansible should fully replace `install.sh`

File: migrations/versions/577b927c0f49_.py
Comment (brassy-endomorph) [inline migrations/versions/577b927c0f49_.py:1]: Migrations shouldn't be part of this PR.

File: pyproject.toml [hunk 1]
```diff
@@ -47,6 +48,7 @@ line_length = 100
 pythonpath = "."
 
 [tool.mypy]
+python_version = "3.11"
```
Comment (brassy-endomorph) [inline pyproject.toml:51]: Why was this added?

File: setup.yml [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+          - redis
+          - redis-server
+          - apt-transport-https
+          - curl
+          - sqlite3
+          - acl
+        update_cache: yes
+        state: latest
+
+    - name: Clone Hush Line repository
```
Comment (brassy-endomorph) [inline setup.yml:32]: This shouldn't be a clone. That implies that we're taking the latest `HEAD` version of the default branch (currently `main`). All the files are local to the repo and should be copied over with `rsync` or something similar.

File: setup.yml [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      git:
+        repo: https://github.com/scidsg/hushline.git
+        dest: /var/www/html/hushline
+        version: ansible
+
+    - name: Install Poetry
```
Comment (brassy-endomorph) [inline setup.yml:38]: This should be installed via `apt`

File: setup.yml [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    - name: Create virtual environment
+      shell: python3 -m venv /var/www/html/hushline/venv
+      args:
+        creates: /var/www/html/hushline/venv
+
+    - name: Configure Poetry to use existing virtual environment
+      shell: |
+        /root/.local/bin/poetry config virtualenvs.create false
+      args:
+        executable: /bin/bash
```
Comment (brassy-endomorph) [inline setup.yml:53]: We don't need to specify `bash` each time.

File: setup.yml [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+    - name: Install dependencies with Poetry
+      shell: |
+        source /var/www/html/hushline/venv/bin/activate && /root/.local/bin/poetry install
+      args:
+        chdir: /var/www/html/hushline
+        executable: /bin/bash
+
+    - name: Perform database migrations
```
Comment (brassy-endomorph) [inline setup.yml:62]: This shouldn't be done in ansible but rather `systemd` via `ExecStartPre`

File: setup.yml [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      shell: |
+        source /var/www/html/hushline/venv/bin/activate && flask db upgrade
+      args:
+        chdir: /var/www/html/hushline
+        executable: /bin/bash
+      become: no
+      become_user: hushlineuser
```
Comment (brassy-endomorph) [inline setup.yml:69]: `become_user` used before user creation

File: setup.yml [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        owner: hushlineuser
+        group: www-data
+
+    - name: Ensure /etc/hushline directory exists
+      file:
+        path: /etc/hushline
+        state: directory
+        owner: hushlineuser
+        group: www-data
+        mode: '0755'
```
Comment (brassy-endomorph) [inline setup.yml:92]: Mode `755` is unsafe as this directory will have sensitive configurations. Should be `0700`

File: vars/secrets/secrets.yml
Comment (brassy-endomorph) [inline vars/secrets/secrets.yml:1]: Secret values for our private instance shouldn't be in the FOSS repo

File: vars/common.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+hushline_user: hushlineuser
```
Comment (brassy-endomorph) [inline vars/common.yml:1]: We have a var `hushline_user` set here but we hardcode the string `hushlineuser` in a bunch of places.

1. We should check that vars are reused and not hardcoded
2. The user should just be `hushline` since the `user` part is implied

File: hushline.nginx.j2 [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    include snippets/ssl-params.conf;
+
+    # Root directory
+    root /var/www/html/hushline;
+
+    # Proxy pass configuration for handling requests
+    location / {
+        proxy_pass http://unix:/var/www/html/hushline/hushline.sock;
+        proxy_set_header Host $host;
+        proxy_set_header X-Real-IP $remote_addr;
```
Comment (brassy-endomorph) [inline hushline.nginx.j2:26]: If we're using CF for production, we don't want to `set` these headers since the real IP would be set to the CF IPs and thus would break all our rate limiting

**Review by brassy-endomorph — CHANGES_REQUESTED** [review]
a few things to start, and a few more comments in the other PR that this should be rebased on top of once that's merged

**Comment by glenn-sorrentino** [comment]
Closing this PR as @brassy-endomorph suggested taking on this work.

File: setup.yml (hunk 3 above)
Comment (glenn-sorrentino) [reply setup.yml:38]: I don't think poetry is available with `apt`. Their documentation suggests that the curl command is the "official" method: https://python-poetry.org/docs/#installing-with-the-official-installer

File: hushline.nginx.j2 (hunk 9 above)
Comment (glenn-sorrentino) [reply hushline.nginx.j2:26]: Resolving as we've moved, at least for now, away from CF.

File: vars/secrets/secrets.yml
Comment (glenn-sorrentino) [reply vars/secrets/secrets.yml:1]: added to gitignore
